types/next-auth/index.d.ts

```diff
@@ -103,7 +103,7 @@ interface JWTOptions {
 
 // TODO: Improve callback typings
 interface Callbacks {
-    signIn?(user: GenericObject, account: GenericObject, profile: GenericObject): Promise<boolean>;
+    signIn?(user: GenericObject, account: GenericObject, profile: GenericObject): Promise<boolean | never | string>;
```

Comment:
I believe `never` is dropped from unions.  What was the intention here?  The thread suggests that you wanted to convey that it might throw, but I think you have assume that about all functions in JS.

Reply:
Yeah, you are probably right. Although, I have another problem.

Looking at the [source code](https://github.com/nextauthjs/next-auth/blob/40e453076e9f008a47916cbdcdc5d6a46b262d54/src/server/routes/callback.js#L70-L84), If am not wrong, it contradicts with what @iaincollins [said](https://github.com/nextauthjs/next-auth/pull/730#issuecomment-701766431). As far as I can tell, returning a string will not actually redirect anywhere. Also, throwing an error like `throw new Error("https://example.com")` will not work, see the source code, `Error`s are caught before `redirect`. So you really must do `throw "https://example.com"`, and I did just that in my code for it to work:
![image](https://user-images.githubusercontent.com/18369201/94987839-8f5eed00-0569-11eb-8f4e-49ce27c2d3c5.png)

But checking it in the `types/next-auth` test file, I get a warning:
![image](https://user-images.githubusercontent.com/18369201/94987880-c7fec680-0569-11eb-81de-448be8d54c5a.png)

I guess it works in my code, but maybe the original source code could handle it better, instead of throwing a string, you could actually return it, which is checked in an `if` statement in the `try` block.



```diff
  try {
    const signInCallbackResponse = await callbacks.signIn(userOrProfile, account, OAuthProfile)
    if (signInCallbackResponse === false) {
      return redirect(`${baseUrl}${basePath}/error?error=AccessDenied`)
+   } else if (typeof signInCallbackResponse === "string") {
+     return redirect(signInCallbackResponse)
+   }
  } catch (error) {
-   if (error instanceof Error) {
      return redirect(`${baseUrl}${basePath}/error?error=${encodeURIComponent(error)}`)
-   } else {
-     return redirect(error)
-   }
  }
```

Reply:
I don't know anything about the underlying library, so it would be better to get confirmation from an area expert.  Having said that, the code you linked clearly doesn't consume any strings that might be returned from `signIn`.  Based solely on that code fragment (the appropriateness of which I can't evaluate), `Promise<string>` appears to be the correct return type.

Reply:
I think @amcasey meant returning `Promise<boolean>`. If throwing a string is an intended feature for `next-auth`, then you can either disable lint for that line, or explicitly cast the string to `any`.